At most 10 `docker run` commands are allowed in parallel

sven-lange-last opened the pull request:
Docker < 1.13.1 has a known problem: if more than 10 containers are created (`docker run`) concurrently, there is a good chance that some of them will fail. See https://github.com/moby/moby/issues/29369

Use a semaphore to make sure that at most 10 `docker run` commands are active the same time.

sven-lange-last commented:
PG 2/2361 running.

markusthoemmes commented on tests/src/test/scala/whisk/core/containerpool/docker/test/DockerClientTests.scala:
```diff
@@ -169,6 +175,101 @@ class DockerClientTests extends FlatSpec with Matchers with StreamLogging with B
+    // The tested code won't reach the mocked executeProcess() and thus, increase runCmdCount,
+    // until at least one Future is successfully completed. For this reason, it takes
+    // some time until the following matcher is successful.
+    eventually(timeout(scaled(Span(5, Seconds)))) { runCmdCount shouldBe 1 }
```
Can this globally be solved by using a `PatienceConfig`?

sven-lange-last replied:
No other tests in class `DockerClientTests` use `eventually` or the `PatienceConfig`yet, so it's safe to introduce a global setting. Done.

markusthoemmes commented on tests/src/test/scala/whisk/core/containerpool/docker/test/DockerClientTests.scala:
```diff
@@ -169,6 +175,101 @@ class DockerClientTests extends FlatSpec with Matchers with StreamLogging with B
+
+    // Cannot assert that the first Docker run always obtains the first container because
+    // the tested code uses Futures so that sequence may differ from test run to test run.
+    val resultContainerIds = mutable.Set[ContainerId]()
```
Mutable? The horror 😱 .  Is that needed here?

sven-lange-last replied:
Replaced by a different approach that only uses immutables.

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerClient.scala:
```diff
@@ -75,25 +77,46 @@ class DockerClient(dockerHost: Option[String] = None)(executionContext: Executio
   }
 
+  protected[containerpool] val maxParallelRuns = 10
+  protected[containerpool] val runSemaphore = new Semaphore( /* permits= */ maxParallelRuns, /* fair= */ true)
```
Can this be only `protected`?

sven-lange-last replied:
Good finding. I'm going to change this.

csantanapr commented:
Should we have an issue open to talk/discuss/track/investigate to update to new version of docker?

If user uses newer version of docker on its own can it have more paralllels running? It needs to configure an ansible setting?

sven-lange-last commented:
@markusthoemmes Fully addressed your review feedback with latest commit. PG 2/2361 still running. Will kick off another PG with latest changes once I have your feedback.

sven-lange-last commented:
Latest Travis build failed due to following problem which is in no way related to this PR:

```
whisk.core.controller.test.ActivationsApiTests > Activations API should get summary activation by namespace and action name FAILED
    org.scalatest.exceptions.TestFailedException: 2 was not equal to 0
        at org.scalatest.MatchersHelper$.indicateFailure(MatchersHelper.scala:340)
        at org.scalatest.Matchers$ShouldMethodHelper$.shouldMatcher(Matchers.scala:6668)
        at org.scalatest.Matchers$AnyShouldWrapper.should(Matchers.scala:6704)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11$$anonfun$apply$30$$anonfun$apply$31.apply(ActivationsApiTests.scala:328)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11$$anonfun$apply$30$$anonfun$apply$31.apply(ActivationsApiTests.scala:325)
        at scala.util.DynamicVariable.withValue(DynamicVariable.scala:58)
        at akka.http.scaladsl.testkit.RouteTest$$anonfun$check$1.apply(RouteTest.scala:57)
        at akka.http.scaladsl.testkit.RouteTest$$anonfun$check$1.apply(RouteTest.scala:57)
        at akka.http.scaladsl.testkit.RouteTestResultComponent$RouteTestResult.$tilde$greater(RouteTestResultComponent.scala:50)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11$$anonfun$apply$30.apply(ActivationsApiTests.scala:325)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11$$anonfun$apply$30.apply(ActivationsApiTests.scala:325)
        at scala.util.Try$.apply(Try.scala:192)
        at whisk.utils.retry$.apply(Retry.scala:44)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11.apply(ActivationsApiTests.scala:324)
        at whisk.core.controller.test.ActivationsApiTests$$anonfun$11.apply(ActivationsApiTests.scala:284)
        at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
        at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
        at org.scalatest.Transformer.apply(Transformer.scala:22)
        at org.scalatest.Transformer.apply(Transformer.scala:20)
        at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
        at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
        at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
        at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
        at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
        at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
        at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
        at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
        at whisk.core.controller.test.ActivationsApiTests.org$scalatest$BeforeAndAfter$$super$runTest(ActivationsApiTests.scala:53)
        at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:203)
        at whisk.core.controller.test.ActivationsApiTests.runTest(ActivationsApiTests.scala:53)
        at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
        at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
        at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
        at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
        at scala.collection.immutable.List.foreach(List.scala:392)
        at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
        at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
        at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
        at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
        at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
        at org.scalatest.Suite$class.run(Suite.scala:1147)
        at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
        at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
        at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
        at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
        at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
        at whisk.core.controller.test.ActivationsApiTests.org$scalatest$BeforeAndAfter$$super$run(ActivationsApiTests.scala:53)
        at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:258)
        at whisk.core.controller.test.ActivationsApiTests.org$scalatest$BeforeAndAfterAll$$super$run(ActivationsApiTests.scala:53)
        at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
        at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
        at whisk.core.controller.test.ActivationsApiTests.run(ActivationsApiTests.scala:53)
```

sven-lange-last commented:
@csantanapr - your questions:

> Should we have an issue open to talk/discuss/track/investigate to update to new version of docker?
>
> If user uses newer version of docker on its own can it have more paralllels running? It needs to configure an ansible setting?

* A [search in the incubator-openwhisk repo](https://github.com/apache/incubator-openwhisk/search?utf8=%E2%9C%93&q=%221.12%22&type=) shows that we require / base on Docker version 1.12.x. At the moment, there is no _real_ support for later or even multiple Docker versions.
* While I think updating the Docker version may bring benefits, IBMs experience is that updating the Docker version implies non-negligible risk of failures and performance degradation. For this reason, this activity means considerable effort and needs a conscious decision.
* As long as we do not officially support multiple Docker versions, we should not spend the effort to introduce an Ansible setting to control the `docker run` command concurrency.

sven-lange-last commented:
PG 2 / 2361: 🔵 

Kicked off PG 4 / 1062 with recent commits.

csantanapr commented:
@sven-lange-last I agree with your assessment 

I still  think we should open an issue to illustrate the work to be done and the effort that would take
If someone outside IBM wants to invest development resources to help in this investigation and performance testing then they will have a blueprint to follow. 

Also for any user that asks when we will support new docker version, well we can explain in the issue that is not as simple as updating a version string in a file :-)

sven-lange-last commented:
PG 4 / 1062 finished with an error in a unit test introduced with this PR: `whisk.core.containerpool.docker.test.DockerClientTests.DockerClient should tolerate docker run errors when limiting the number of concurrent docker run invocations`. i need to investigate further.

sven-lange-last commented:
Found that the problem in PG 4 / 1062 reported above was caused by sequencing problems in the test. Provided a commit that addresses these problems.

Initiated PG3 / 1393.

sven-lange-last commented:
PG3 / 1393 🔵
